tools/ipmonitor/main.cpp

```diff
@@ -0,0 +1,428 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#include <QCommandLineParser>
+#include <QCoreApplication>
+#include <QFile>
+#include <QHash>
+#include <QJsonArray>
+#include <QJsonDocument>
+#include <QJsonObject>
+#include <QJsonValue>
+#include <QTextStream>
+#include <QScopeGuard>
+
+#include <net/ethernet.h>
+#ifdef MVPN_LINUX
+#  include <netinet/ether.h>
+#elif MVPN_MACOS
+#  include <netinet/if_ether.h>
+#endif
+#include <pcap.h>
+#include <sys/ioctl.h>
+#include <stdio.h>
+#include <unistd.h>
+
+#include "rfc/rfc1918.h"
+
+/* IP header. See https://www.tcpdump.org/pcap.html */
+struct sniff_ip {
+  u_char ip_vhl;          /* version << 4 | header length >> 2 */
+  u_char ip_tos;          /* type of service */
+  u_short ip_len;         /* total length */
+  u_short ip_id;          /* identification */
+  u_short ip_off;         /* fragment offset field */
+#define IP_RF 0x8000      /* reserved fragment flag */
+#define IP_DF 0x4000      /* don't fragment flag */
+#define IP_MF 0x2000      /* more fragments flag */
+#define IP_OFFMASK 0x1fff /* mask for fragmenting bits */
+  u_char ip_ttl;          /* time to live */
+  u_char ip_p;            /* protocol */
+  u_short ip_sum;         /* checksum */
+  in_addr ip_src, ip_dst; /* source and dest address */
+#define IP_HL(ip) (((ip)->ip_vhl) & 0x0f)
+#define IP_V(ip) (((ip)->ip_vhl) >> 4)
+};
+
+class O final : public QTextStream {
```

Comment:
I dislike the single-character name for this class, could you please name it to something that reflects its purpose? `Output` or `Console` or something?